Fix unsigned conversion bug. And switchuvm.

phf opened the pull request:
Since readi() returns -1 for errors, checking with < against an unsigned value is inadvisable. Checking with != works as intended however. Thanks to @nwf for double-checking.

nubok commented:
I ask you whether it wouldn't be better to change 

`if(readi(ip, (char*)&elf, 0, sizeof(elf)) < sizeof(elf))`

to

`if(readi(ip, (char*)&elf, 0, sizeof(elf)) < (int) sizeof(elf))`

instead.

nwf commented:
I'd argue for != for consistency with the readi() call a few lines down, if for no other reason.  Given that there is precisely one bit pattern -- which happens to be the encoding of sizeof(elf) -- that should pass this test, what's the advantage of "< (int)"?

nubok commented:
"When comparing signed with unsigned, the compiler converts the signed value to unsigned." (cf. http://stackoverflow.com/a/5416498/497193). Such an implicit conversion obscurs the clarity of the code in my opinion (put hand up if you were 100% sure about the conversion rules). Thus I personally prefer a code style where as few implicit conversions as possible happen.

phf commented:
I can maybe get behind the explicit cast, although I don't think that's what the rest of xv6 looks like and we might want to stick to existing style. However, the < still doesn't make sense: We want to read *exactly* the ELF header, no more, no less. Furthermore there is no scenario under which readi should return more bytes then we asked for. (Unless I misunderstand readi on a very deep level.)

phf commented:
(Also note that you're not making an implicit conversion explicit, you're just applying a different conversion.)

phf commented:
Sorry, just added another commit here that fixes what I believe to be a long-standing bug in switchuvm(). Alas the bug didn't ever break a test case, but if you read the details I think you'll agree that it is indeed a bug. I also added a few more checks and moved those outside the pushcli/popcli pair. Happy holidays!

kaashoek commented:
Applied, thanks.
